JoomGalleryfriends/JG4-dev
# FileManager service

## What is this service for

The FileManager service will be used to handle the files and folders created in the filesystem. I offers methods to perform processes (create, delete, check, move) on image files and folders. The file structure is created based on...

- ... the current available image types (ImagetypeTable)
- ... the parameters from the configuration set of the current user (ConfigService)
- ... the chosen filesystem (FilesystemService)
- ... the chosen image processor (IMGtoolsService)

## How to use this service

You can load the service anywhere in Joomla by doing

```
$component = Factory::getApplication()->bootComponent('com_joomgallery');  // boot JoomGallery component
$component->createFileManager();   // instantiate service
$file_manager = $component->getFileManager();   // load the service into your scope
$file_manager ->createImages('image.jpg', 'new_filename.jpg', 1);  // create image files in category with id=1
$file_manager ->deleteImages(22);   // delete image files with image id=22
```

## How to test this service

Make sure "Debug System" is "Yes" in Joomla Global configuration.
Open the file `/administrator/com_joomgallery/src/View/Faulties/HtmlView.php` and insert the following code on line 62 after `throw new \Exception(implode("\n", $errors)); }`

```
    // Test FileManager
    $this->component->createConfig();
    $this->component->createFileManager();
    $manager = $this->component->getFileManager();
    $src     = \Joomla\CMS\Filesystem\Path::clean(JPATH_ROOT.'/images/<YOUR_IMAGE.jpg>');
    $settings = array(  'filename'       => '<YOUR_FILENAME.jpg>',
                        'imgid'          => <IMGID>,
                        'catid'          => <CATID>,
                        'parent_catid'   => <PARENT_CATID>,
                        'cat_alias'      => '<YOUR_CATEGORY>');

    // Create Category
    //$manager->createCategory($settings['cat_alias'], $settings['parent_catid']);

    // Create Image
    // $manager->createImages($src, $settings['filename'], $settings['catid']);

    // Check Image
    // $img_info = $manager->checkImages($settings['filename']);
    // dump($img_info);

    // Get Imagetype path of existing image
    // $img_path = $manager->getImgPath('original', $settings['imgid']);
    // dump($img_path);

    // Copy Image
    // $manager->copyImages($settings['filename'], 5);

    // Move Image
    // $manager->moveImages($settings['filename'], 5);

    // Check Category
    // $cat_info = $manager->checkCategory($settings['cat_alias']);
    // dump($cat_info);

    // Get Category path of existing image
    // $cat_path = $manager->getCatPath('original', $settings['catid']);
    // dump($cat_path);

    // Move Category
    // $manager->moveCategory($settings['catid'], 5);

    // Copy Category
    // $manager->copyCategory($settings['catid'], 5);

    // Delete Category
    // $manager->deleteCategory($settings['cat_alias']);

    // Delete Image
    // $manager->deleteImages($settings['filename']);
```

Create a category using the category manager and give it the alias=<YOUR_CATEGORY>. You can nest this category as you like below any other category already created. Note the id of the direct parent category of the new created category (<PARENT_CATID>).

Category Manager:
![grafik](https://user-images.githubusercontent.com/39154009/177053516-50cdae45-ebbb-4b02-9af4-62b3b7c53163.png)

---

Create an image using the image manager. Select the category you just created (<YOUR_CATEGORY>). Go into the database and edit your just created image record in the table #__joomgallery. In the row 'filename' write the filename as which you want the created files to be stored (<YOUR_FILENAME.jpg>).

Image Manager:
![grafik](https://user-images.githubusercontent.com/39154009/177418107-a80655a2-6d5b-45ab-9490-910f4d76d63a.png)

Database (#__joomgallery)
![grafik](https://user-images.githubusercontent.com/39154009/177418329-48483ada-1605-4e6d-9000-547de8503a19.png)

Navigate to Components->JoomGallery->Maintenance Manager to see the FileManager service running...

## What to test
### Different configuration settings
General Settings >> Image Processing >> Image processing (static images)

### Different source images and destination categories
- `<YOUR_IMAGE.jpg>` (Different source images)
- `<YOUR_FILENAME.jpg>` (Different filenames of the image)
- `<YOUR_CATEGORY>` (Different categories - alias)
- `<CATID>` (Different categories - id)
- `<PARENT_CATID>` (Different parent categories - id)

The service will create a folder structure and images in such a way as you defined it in your image entry and your category entries in the database (including nesting).

### Different processing methods
Test the different methods from the service
Available methods:
- createImages($source, $filename, $cat=2)
- deleteImages($img)
- checkImages($img)
- moveImages($img, $dest, $copy=false)
- copyImages($img, $dest)
- createCategory($foldername, $parent=1)
- deleteCategory($cat, $del_images=false)
- checkCategory($cat)
- moveCategory($cat, $dest, $copy=false)
- copyCategory($cat, $dest)
- getImgPath($type, $img, $catid=false, $filename=false, $root=0)
- getCatPath($cat, $type=false, $parent=false, $alias=false, $root=0)

more infos: [FileManager interface](https://github.com/Elfangor93/JG4-dev/blob/ImageMgr/administrator/com_joomgallery/src/Service/FileManager/FileManagerInterface.php)
